File: src/lang/clj/clojure/set.clj
;; clojure.set — Phase 6.16.b-1 (.clj migration).
;;
;; Loaded by `src/lang/bootstrap.zig::loadCore` per ADR-0032 multi-file
;; FILES table. The Group A + B vars (`union` / `intersection` /
;; `difference` / `subset?` / `superset?` / `rename-keys` /
;; `map-invert`) are pure-Clojure Pattern A defns per ADR-0033 D3 + v5
;; §8.2. Each composes `reduce` / `conj` / `disj` / `contains?` /
;; `every?` / `assoc` / `dissoc` / `get` / `count` from rt/ — visible
;; unqualified here because `evalInNs` refers rt/ into the entered ns
;; (commit 6.16.b-1 + ADR-0035 in Phase 6.16.b-4 codifies this as a
;; proper `(ns ...)` macro).
;;
;; **Variadic via [& sets] + internal arity discrimination**: union /
;; intersection / difference accept 0/1/2/3+ args using a single
;; rest-arg `fn*` form (no multi-arity dispatch needed). This sidesteps
;; D-070 (multi-arity `fn*`) for these three vars.
;;
;; Group C (`select` / `project` / `index` / `rename` / `join`) lands
;; at 6.16.b-3 after D-061 (`#{}` reader literal) + D-059 (map-literal
;; analyzer) infra ships in 6.16.b-2.

(ns clojure.set (:refer-clojure))

(def union
  (fn* [& sets]
    (if (= 0 (count sets))
      (hash-set)
      (reduce (fn* [acc s] (reduce conj acc s))
              (first sets)
              (rest sets)))))

(def intersection
  (fn* [& sets]
    (if (= 0 (count sets))
      nil
      (if (= 1 (count sets))
        (first sets)
        (reduce (fn* [s1 s2]
                  (reduce (fn* [acc x]
                            (if (contains? s2 x) acc (disj acc x)))
                          s1
                          s1))
                (first sets)
                (rest sets))))))

(def difference
  (fn* [& sets]
    (if (= 0 (count sets))
      nil
      (if (= 1 (count sets))
        (first sets)
        (reduce (fn* [s1 s2] (reduce disj s1 s2))
                (first sets)
                (rest sets))))))

(def subset?
  (fn* [s1 s2]
    (if (<= (count s1) (count s2))
      (every? (fn* [x] (contains? s2 x)) s1)
      false)))

(def superset?
  (fn* [s1 s2] (subset? s2 s1)))

;; `(rename-keys m kmap)` — rebuild m by replacing each old key in
;; kmap with its new-key partner. Skips entries whose old key is not
;; in m (matches JVM). The `(nth kv 0/1)` destructure substitutes
;; for vector binding inside `let*`.
(def rename-keys
  (fn* [m kmap]
    (reduce (fn* [acc kv]
              ;; nth over the map-entry vector is the Pattern-A finished form
              ;; (mirrors map-invert below); `let*` is the primitive special
              ;; form and never gains destructure — that is `let`'s job, which
              ;; this bootstrap-layer file deliberately avoids.
              (let* [old (nth kv 0)
                     new-k (nth kv 1)]
                (if (contains? m old)
                  (assoc (dissoc acc old) new-k (get m old))
                  acc)))
            m
            kmap)))

;; `(map-invert m)` — swap keys and values. Matches JVM's transient
;; reduce-kv shape (D-074 cycle 3 discharged the PROVISIONAL marker).
(def map-invert
  (fn* [m]
    (persistent!
      (reduce (fn* [acc kv]
                (assoc! acc (nth kv 1) (nth kv 0)))
              (transient (hash-map))
              m))))

;; ----------------------------------------------------------------
;; Group C — relational ops (Phase 6.16.b-3). Sits on top of D-061
;; (#{} reader) + D-059 (map literal as Value) infra landed at
;; 6.16.b-2. select-keys / merge / set helpers come from core.clj.
;;
;; project / rename preserve source metadata via `with-meta` + `meta`
;; (value-metadata system landed). join ships the full 1/2/3-arity
;; surface, including the 3-arity `[xrel yrel km]` key-mapping form
;; (multi-arity `fn*` per ADR-0041 / D-070 discharge).
;; ----------------------------------------------------------------

;; `(select pred xset)` — return the subset of `xset` whose
;; elements satisfy `pred`.
(def select
  (fn* [pred xset]
    (reduce (fn* [s k] (if (pred k) s (disj s k))) xset xset)))

;; `(project xrel ks)` — return a rel containing only the keys in
;; `ks` for each map in `xrel`.
(def project
  (fn* [xrel ks]
    (with-meta (set (map (fn* [m] (select-keys m ks)) xrel)) (meta xrel))))

;; `(rename xrel kmap)` — return a rel with the keys in each map
;; renamed per kmap.
(def rename
  (fn* [xrel kmap]
    (with-meta (set (map (fn* [m] (rename-keys m kmap)) xrel)) (meta xrel))))

;; `(index xrel ks)` — return a map of (selected-keys → set-of-maps).
(def index
  (fn* [xrel ks]
    (reduce (fn* [m x]
              (let* [ik (select-keys x ks)]
                (assoc m ik (conj (get m ik #{}) x))))
            {}
            xrel)))

;; `(join xrel yrel)` — natural join on the common keys.
;; `(join xrel yrel km)` — arbitrary key mapping; `km` maps keys of
;; `xrel` to the corresponding keys of `yrel`. Multi-arity landed at
;; row 7.8 cycle 4 per ADR-0041 (D-070 discharge).
(def join
  (fn*
    ([xrel yrel]
     (if (and (seq xrel) (seq yrel))
       (let* [ks (intersection (set (keys (first xrel)))
                               (set (keys (first yrel))))
              smaller? (<= (count xrel) (count yrel))
              r (if smaller? xrel yrel)
              s (if smaller? yrel xrel)
              idx (index r ks)]
         (reduce (fn* [ret x]
                   (let* [found (get idx (select-keys x ks))]
                     (if found
                       (reduce (fn* [acc m] (conj acc (merge m x))) ret found)
                       ret)))
                 #{}
                 s))
       #{}))
    ([xrel yrel km]
     (let* [smaller? (<= (count xrel) (count yrel))
            r (if smaller? xrel yrel)
            s (if smaller? yrel xrel)
            k (if smaller? (map-invert km) km)
            idx (index r (vals k))]
       (reduce (fn* [ret x]
                 (let* [found (get idx (rename-keys (select-keys x (keys k)) k))]
                   (if found
                     (reduce (fn* [acc m] (conj acc (merge m x))) ret found)
                     ret)))
               #{}
               s)))))

File: src/lang/primitive/set.zig
// SPDX-License-Identifier: EPL-2.0
//! `clojure.set/` namespace surface + the set/map constructors.
//!
//! The Group A + B vars (`union` / `intersection` / `difference` /
//! `subset?` / `superset?` / `rename-keys` / `map-invert`) live as
//! pure-Clojure Pattern A defns in `src/lang/clj/clojure/set.clj`
//! per ADR-0033 D3 + v5 §8.2. This file ships only the variadic
//! constructors:
//!
//! - `hash-set` — `(hash-set & xs)`. (The `#{...}` reader literal has
//!   its own `.set` Form variant, so it does not route through here.)
//! - `hash-map` / `array-map` — `(hash-map & kvs)`. (The `{...}`
//!   literal likewise has its own Form variant.)
//!
//! Group C (`select` / `project` / `index` / `rename` / `join`,
//! relational ops over sets-of-maps) is not yet implemented; it is
//! tracked by D-061.

const Value = @import("../../runtime/value/value.zig").Value;
const Runtime = @import("../../runtime/runtime.zig").Runtime;
const env_mod = @import("../../runtime/env.zig");
const Env = env_mod.Env;
const error_mod = @import("../../runtime/error/info.zig");
const SourceLocation = error_mod.SourceLocation;
const error_catalog = @import("../../runtime/error/catalog.zig");
const dispatch = @import("../../runtime/dispatch.zig");
const set_collection = @import("../../runtime/collection/set.zig");
const map_collection = @import("../../runtime/collection/map.zig");

/// `(hash-set & xs)` — construct a set from variadic args. Empty
/// arg list returns the empty-set singleton. Each arg is conj-ed
/// in order (idempotent — duplicates collapse).
pub fn hashSet(rt: *Runtime, env: *Env, args: []const Value, loc: SourceLocation) anyerror!Value {
    _ = env;
    _ = loc;
    var s = set_collection.empty();
    for (args) |a| s = try set_collection.conj(rt, s, a);
    return s;
}

/// `(hash-map & kvs)` — construct a map from variadic key/value pairs.
/// Odd argument count raises `map_literal_arity_odd` (matches the
/// JVM `IllegalArgumentException`). Empty arg list returns the
/// empty-map singleton.
pub fn hashMap(rt: *Runtime, env: *Env, args: []const Value, loc: SourceLocation) anyerror!Value {
    _ = env;
    if (args.len % 2 != 0)
        return error_catalog.raise(.map_literal_arity_odd, loc, .{});
    var m = map_collection.empty();
    var i: usize = 0;
    while (i < args.len) : (i += 2) {
        m = try map_collection.assoc(rt, m, args[i], args[i + 1]);
    }
    return m;
}

// --- registration ---

const Entry = struct {
    name: []const u8,
    f: dispatch.BuiltinFn,
};

const RT_ENTRIES = [_]Entry{
    .{ .name = "hash-set", .f = &hashSet },
    .{ .name = "hash-map", .f = &hashMap },
    // array-map: a cljw map starts array-backed (insertion-ordered) for ≤8
    // entries, so for the realistic small-map surface array-map ≡ hash-map's
    // output. (Residual: clj's array-map never hash-promotes; cljw promotes
    // past 8 entries — a >8-entry array-map is rare.)
    .{ .name = "array-map", .f = &hashMap },
};

/// Register `hash-set` / `hash-map` into `rt/` (so they are user-
/// callable unqualified after `(refer 'rt)` into `user/`) and ensure
/// the `clojure.set` namespace exists for the .clj loader to enter.
/// The Group A + B vars themselves are registered by evaluating
/// `src/lang/clj/clojure/set.clj` at bootstrap.
pub fn register(env: *Env) !void {
    const core_ns = env.findNs("clojure.core") orelse return error.ClojureCoreNamespaceMissing;
    for (RT_ENTRIES) |it| {
        _ = try env.intern(core_ns, it.name, Value.initBuiltinFn(it.f), null);
    }
    _ = try env.findOrCreateNs("clojure.set");
}
